pysmt/optimization/optimizer.py

```diff
@@ -151,34 +202,190 @@ def optimize(self, cost_function, strategy='ub',
 
         `step_size` the minimum reolution for finding a solution. The
         optimum will be found in the proximity of `step_size`
-
         """
+        rt = None, None
+        if goal.is_simple_goal():
+            rt = self._simple_optimize(goal, strategy,
+                             feasible_solution_callback,
+                             step_size, **kwargs)
+        else:
+            raise GoalNotSupportedError("ExternalOptimizerMixin", goal)
+        return rt
+
+
+
+    def _simple_optimize(self, goal, strategy,
+                      feasible_solution_callback,
+                      step_size, **kwargs):
+        rt = None
+        if goal.is_maximization_goal():
+            rt = self._optimize_max(goal, strategy,
+                             feasible_solution_callback,
+                             step_size, **kwargs)
+        elif goal.is_minimization_goal():
+            rt = self._optimize_min(goal, strategy,
+                             feasible_solution_callback,
+                             step_size, **kwargs)
+        elif goal.is_minmax_goal():
+            rt = self._optimize_minmax(goal, strategy,
+                                    feasible_solution_callback,
+                                    step_size, **kwargs)
+        elif goal.is_maxmin_goal():
+            rt = self._optimize_maxmin(goal, strategy,
+                                    feasible_solution_callback,
+                                    step_size, **kwargs)
+        else:
+            raise GoalNotSupportedError("ExternalOptimizerMixin", goal)
+        return rt
+
+    def _optimize_minmax(self, goal, strategy,
+                 feasible_solution_callback,
+                 step_size, **kwargs):
+
+        part_res = []
+
+        for term in goal.terms():
+            self.push()
+            part_res.append(self._optimize_min(MinimizationGoal(term),
+                                          strategy, feasible_solution_callback,
+                                          step_size, **kwargs)[1]
+                            )
+            self.pop()
+        rt = None
+        self.push()
+        t = goal.terms()[0]
+        le = self._get_le(t)
+        _or = self._get_or(t)
+
+        symbol_type = self._get_symbol_type(t)
+
+        _private_symbol = Symbol("_private_symbol", symbol_type)
+        f = []
+        for res in part_res:
+            f.append(le(_private_symbol, res))
+
+        self.add_assertion(_or(f))
```

Comment:
Nope! you cannot have `_private_symbol` to be smaller than at least one element, you need it to be smaller than all the elements and to be equal to al least one of them to be the minimum....